python/ql/src/semmle/python/frameworks/PEP249.qll

```diff
@@ -7,20 +7,23 @@ private import python
 private import semmle.python.dataflow.new.DataFlow
 private import semmle.python.dataflow.new.RemoteFlowSources
 private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
 
-/** A module implementing PEP 249. Extend this class for implementations. */
-abstract class PEP249Module extends DataFlow::Node { }
+/**
+ * A module implementing PEP 249. Extend this class for implementations.
+ *
+ * DEPRECATED: Extend `PEP249ModuleApiNode` instead.
+ */
+abstract deprecated class PEP249Module extends DataFlow::Node { }
 
-/** Gets a reference to a connect call. */
-private DataFlow::Node connect(DataFlow::TypeTracker t) {
-  t.startInAttr("connect") and
-  result instanceof PEP249Module
-  or
-  exists(DataFlow::TypeTracker t2 | result = connect(t2).track(t2, t))
-}
+/**
+ * An abstract class encompassing API graph nodes that implement PEP 249.
+ * Extend this class for implementations.
+ */
+abstract class PEP249ModuleApiNode extends API::Node { }
```

Comment:
Could we not just add the `toString` method here, and thus not require it in every model?

Reply:
> I am slightly worried about the potential foot-gun of no toString-values during debugging.

Just to be clear, all of the relevant nodes have `toString` representations at the moment. The only time this becomes a potential foot-gun is when we're adding new node types to `API::Impl::TApiNode`. In this case, we might forget to add `toString`, though I think it is unlikely.

> Could we not just add the toString method here, and thus not require it in every model?

We could, but what should be the string returned? All nodes already have appropriate `toString`s... I guess we could prefix them with the fact that they're `PEP249ModuleApiNode`s, but I'm not sure that actually adds value.

Maybe I'm doing something odd in terms of inheritance here? :thinking:

Reply:
Would it not be possible to add a test that fails when any API::Impl::TApiNode class does not have a toString method?

Reply:
That's... not a terrible idea. We can't test on `TApiNode` directly, as it's hidden inside a `private` module, but we _could_ do
```ql
select count(API::Node n | not exists(n.toString()))
```
in a test (with expected output `0`) and that should at least warn us if this changes.